src/sm/sm-common.c

```diff
@@ -385,29 +383,32 @@ sm_encrypt_des_ecb3(unsigned char *key, unsigned char *data, int data_len,
 	for (ii=0; ii<data_len; ii+=8)
 		DES_ecb2_encrypt( (sm_des_cblock *)(data + ii),
 				(sm_des_cblock *)(*out + ii), &ks, &ks2, DES_ENCRYPT);
+
+	return 0;
 #else
 	cctx = EVP_CIPHER_CTX_new();
 	if (!EVP_EncryptInit_ex2(cctx, EVP_des_ede_ecb(), key, NULL, NULL)) {
-		EVP_CIPHER_CTX_free(cctx);
-		return SC_ERROR_INTERNAL;
+		goto err;
 	}
 	/* Disable padding, otherwise it will fail to decrypt non-padded inputs */
 	EVP_CIPHER_CTX_set_padding(cctx, 0);
 	if (!EVP_EncryptUpdate(cctx, *out, &tmplen, data, data_len)) {
-		EVP_CIPHER_CTX_free(cctx);
-		return SC_ERROR_INTERNAL;
+		goto err;
 	}
 	*out_len = tmplen;
 
 	if (!EVP_EncryptFinal_ex(cctx, *out + *out_len, &tmplen)) {
-		EVP_CIPHER_CTX_free(cctx);
-		return SC_ERROR_INTERNAL;
+		goto err;
 	}
 	*out_len += tmplen;
 	EVP_CIPHER_CTX_free(cctx);
-#endif
-
 	return 0;
+
+err:
+	EVP_CIPHER_CTX_free(cctx);
```

Comment:
please check with `if (cctx)`

Reply:
Unfortunately, the manual page does not specify this, but OpenSSL API is getting sane semantics and NULL arguments are ok for the free functions:
https://github.com/openssl/openssl/blob/f5b06306b7727f830407c639f8a7bf1698f8aaa5/crypto/evp/evp_enc.c#L84

Reply:
Nice to know; is this a general approach for all APIs? Historically, they typically didn't check any input parameters...

Reply:
I am not aware of specific OpenSSL guidelines, but I think it follows the least surprise principle as the `free()` ignores the `NULL` pointers and people are (finally) getting use to that too as it simplifies a lot of needless null checks on cleanup.